### PR TITLE
My_Enumerable_Methods

### PR DESCRIPTION
**Pull request by Miguelus373**
### What this PR does
- Add methods to the enumerable module
- These methods mirror the behavior of the built-in ones

### CODE REVIEW

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
Hello Team :wave:
Status: Required Changes :recycle:

A great job has done by you so far! :star:

Now this will be long feedback, but the purpose is to show you all the details that you have to implement in one shot.
Right below, you will see a lot of bullet points. 
They are all the same things, that are already written in the documentation.

Documentations?
Here:
- https://ruby-doc.org/core-2.6.5/Enumerable.html
- https://ruby-doc.org/core-2.6.5/Enumerator.html

These will show you the correct way of doing things. For every method on that page, you can see the inputs and outputs. And our own methods should act exactly the same.

One way is to look up from that document and another way is to follow the explanations that I wrote down :arrow_down:
These explanations are just the same things on that document page. I just humanized the instructions by converting them into plain English.

For instance,
### my_each
:camera: **Screenshot: What does the documentation say about each method**
each
:orange_circle:  This is the heading of the "**each**" section in the documentation. Which shows the inputs. You can see that documentation says these things =>

1. It takes a block and returns an object.
2. It returns an enumerator if it doesn't take any input.
3. It takes arguments and it takes a block and returns an object.
4. It takes an argument and returns an enumerator.

:camera: **Screenshot: We also have examples for "each" in the document**
each examples
You can see the examples for every input here.

:teacher: And this is the pure English version (Also adding the edge cases) =>

- returns an Enumerator if no block is given
- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- when a block is given when self is a range behaves like  calls the given block once for each element in self
- when a block is given when self is a hash behaves like  calls the given block once for each element in self

# And this is how we learn to read the documentation! :page_facing_up: :+1:  :1st_place_medal:  :100:

#### After this point I am writing to you the requirements of your own project, please read one by one, and be sure that your own methods are capable of doing these bullet points...

:bulb: **Example**
```
For some of the bullet points that I have chosen randomly you will find an example in this context.
```

## my_each
- returns an Enumerator if no block is given
- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self

:bulb: **Example**
```ruby
(1..5).my_each(&block) should be equal to (1..5).each(&block)
# my_each should return the range itself
```
- when a block is given when self is a range behaves like #each calls the given block once for each element in self

## my_each_with_index
- returns an enumerator if no block is given
- when a block is given when self is an array returns the array itself after calling the given block once for each element in self

:bulb: **Example**
```ruby
array.my_each_with_index(&block) should be equal to array.each_with_index(&block)
# my_each_with_index should return the array itself
```
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

## my_select
- returns an array containing all elements of enum for which the given block returns a true value::range
- returns an enumerator if no block is given
:bulb: **Example**
```ruby
array.my_select #=> should return an Enumerator
```

## my_all?
- returns false if the block returns false or nil::range
- when no block or argument is given returns false when one of the collection members is false or nil
- when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
:bulb: **Example**
```ruby
[1, 2, 3].my_all?(Integer) #=> true
[1, 2, 3, 'string'].my_all?(Integer) #=> false
```
- when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- when a class is passed as an argument returns false if any of the collection is not a member of such class
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
:bulb: **Example**
```ruby
['Cat', 'Bark', 'Dark'].my_all?(/a/) #=> true
['Cat', 'Bark', 'Dark', 'Sun'].my_all?(/a/) #=> false
```
- when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

## my_any?
- returns false if the block ever returns a value false or nil::range
- when no block or argument is given returns false if at least one of the collection is not true
:bulb: **Example** TRICKY, work on the technical terms: TRUTHY && FALSY
```ruby
[false, false, true].my_any? #=> true
[false, false, false, nil].my_any? #=> false
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- when a Regex is passed as an argument returns true if any of the collection matches the Regex
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
:bulb: **Example**
```ruby
['Ankara', 'Gelibolu'].my_any?(/a/) #=> true
['There', 'sun'].my_any?(/a/) #=> false
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

## my_none?
:bulb: **HINT** This is connectable with "my_any?"
- returns false if the block ever returns true for all elements::range
- when no block or argument is given returns true only if none of the collection members is true
- when a class is passed as an argument returns true if none of the collection is a member of such class
- when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- when a class is passed as an argument returns false if any of the collection is a member of such class
- when a Regex is passed as an argument returns true if none of the collection matches the Regex
- when a Regex is passed as an argument returns false if any of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

## my_count
- returns the number of items in enum through enumeration
:bulb: **Example**
```ruby
['Ankara', 'Gelibolu'].my_count # => returns 2
```
- returns the number of items in enum through enumeration::range

## my_map
- returns a new array with the results of running a given block once for every element in enum::range
- returns an Enumerator if no block is given

## my_inject
`Know the inject well and you will rule the Ruby! - a wise man said`
- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
:bulb: **Example**
```ruby
[1, 2, 3, 4, 5].my_inject(:+) # => returns the sum of all the values (15)
[1, 2, 3, 4, 5].my_inject(:*) # => returns the chain multiplication (120)
```
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
:bulb: **Example** I think this is very :cool: ...
```ruby
[1, 2, 3, 4, 5].my_inject(2, :+) # => returns the sum of all the values plus the initial one (17)
[1, 2, 3, 4, 5].my_inject(2, :*) # => returns guess what? yess! (240) :+1: 
```
- when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

## multiply_els
- accepts an array as an argument and multiplies all the elements of the array together using #my_inject

### That's all.

Hey team :wave: I know it seems a bit too much. But this is the core of the Ruby. After improving yourself on these methods, imagine how comfortable you will feel about Ruby!

Please, let me know if you need anything else, via Slack.

With all my respect :pray: ...

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  #       return puts 'Enumerator #{self}:my_map'
+  #     else
+  #       new_arr = []
+  #       if self.is_a? Array
+  #         for i in 0...(self.length)
+  #             new_arr.push(yield(self[i]))
+  #         end
+  #       elsif self.is_a? Hash
+  #         key = self.keys
+  #         value = self.values
```
Comment (kubilaycaglayan) [inline enumerable_methods.rb]: Thank you for being thoughtful :pray: 
You don't have to do that unless it is specified, let's delete this commented out code from our project :blush:

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
Hello Team :wave: 
Status: Changes Required :recycle: 

Congratulations :clap: 
You have made amazing progress in this hard project! :+1:  :+1: 

Seriously... :1st_place_medal: 

What is left is just a couple of requirements and I will be giving you the hints :arrow_down: 

## my_each
- returns an Enumerator if no block is given
 💡 Check the following link. At the very beginning of the page you will find something useful.
 https://ruby-doc.org/core-2.6.5/Enumerator.html
Ummm... :eyes:  something about... :thought_balloon:  :thinking:  **to_enum**
## my_each_with_index
- returns an enumerator if no block is given
## my_select
- returns an enumerator if no block is given
## my_all?
- returns false if the block returns false or nil::range
 💡 HINT: You are using "empty?" to check. But it doesn't work for ::range type.
## my_any?
- returns false if the block ever returns a value false or nil::range
## my_none?
- returns false if the block ever returns true for all elements::range
## my_count
- returns the number of items in enum through enumeration
- returns the number of items in enum through enumeration::range
## my_map
- returns an Enumerator if no block is given

With all my respect... :pray:

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+      (0...(to_array.length)).each do |i|
+        count += 1 if compare == to_array[i]
+      end
+    else
+      (0...(to_array.length)).each do |i|
+        count += 1 if compare == to_array[i]
+      end
+    end
+    count
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: ### My_count

- [x] When `my_count` is called on a range, it should return the number of items in enum. 

- [x] When `my_count` is called on an array, it should return the number of items in enum through enumeration. 

* Yours is currently returning **`nil`** for both of these test cases.

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi @Miguelus373  and @SunnySparks ! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with your methods so far! You are almost done👏 👏 👏  There is just one method that needs to be fixed before moving on.
****

### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

**Review by kubilaycaglayan — APPROVED** [review]
Hey Team :wave: 
Status: Project Approved :green_circle: 

Great job!
You have just mastered the **Ruby enumerable module!** :1234: 

With all my respect :pray:

**Pull request merged**